cmd/restic/cmd_prune.go

```diff
@@ -87,229 +136,433 @@ func runPrune(gopts GlobalOptions) error {
+	Verbosef("total data files:        %10d\n\n",
+		stats.packs.mixed+stats.packs.tree+stats.packs.data+uint(len(removePacksFirst)))
+
+	Verbosef("small data files:       %10d\n", stats.packs.small)
+	Verbosef("totally used data files: %10d\n", stats.packs.used)
+	Verbosef("partly used data files: %10d\n", stats.packs.partlyUsed)
+	Verbosef("unused data files:      %10d\n\n", stats.packs.unused)
+
+	Verbosef("to keep:   %10d data files\n", stats.packs.keep)
+	Verbosef("to repack: %10d data files\n", len(repackPacks))
+	Verbosef("to delete: %10d data files\n", len(removePacks))
+	Verbosef("to delete: %10d unreferenced data files\n\n", len(removePacksFirst))
+
+	if opts.DryRun {
+		if !gopts.JSON && gopts.verbosity >= 2 {
+			Printf("Would have removed the following unreferenced data files:\n%v\n\n", removePacksFirst)
+			Printf("Would have repacked and removed the following data files:\n%v\n\n", repackPacks)
+			Printf("Would have removed the following no longer used data files:\n%v\n\n", removePacks)
 		}
+		// Always quit here if DryRun was set!
+		return nil
+	}
 
-		rewritePacks.Delete(packID)
+	// delete obsolete index files (index files that have already been superseded)
+	obsoleteIndexes := (repo.Index()).(*repository.MasterIndex).Obsolete()
+	if len(obsoleteIndexes) != 0 {
+		Verbosef("deleting unused index files...\n")
+		DeleteFiles(gopts, true, repo, obsoleteIndexes, restic.IndexFile)
 	}
 
-	Verbosef("will delete %d packs and rewrite %d packs, this frees %s\n",
-		len(removePacks), len(rewritePacks), formatBytes(uint64(removeBytes)))
+	// unreferenced packs can be safely deleted first
+	if len(removePacksFirst) != 0 {
+		Verbosef("deleting unreferenced data files...\n")
+		DeleteFiles(gopts, true, repo, removePacksFirst, restic.DataFile)
+	}
 
-	var obsoletePacks restic.IDSet
-	if len(rewritePacks) != 0 {
-		bar = newProgressMax(!gopts.Quiet, uint64(len(rewritePacks)), "packs rewritten")
+	if len(repackPacks) != 0 {
+		Verbosef("repacking data files...\n")
+		bar := newProgressMax(!gopts.Quiet, uint64(len(repackPacks)), "data files repacked")
 		bar.Start()
-		obsoletePacks, err = repository.Repack(ctx, repo, rewritePacks, usedBlobs, bar)
+		// TODO in Repack:  - Parallelize repacking
+		//                  - Make use of blobs stored multiple times (e.g. if SHA doesn't match)
+		_, err := repository.Repack(ctx, repo, repackPacks, keepBlobs, bar)
+		bar.Done()
 		if err != nil {
 			return err
 		}
-		bar.Done()
+		// Also remove repacked packs
+		removePacks.Merge(repackPacks)
 	}
 
-	removePacks.Merge(obsoletePacks)
+	if len(removePacks) != 0 {
+		Verbosef("rebuilding index files...\n")
 
-	if err = rebuildIndex(ctx, repo, removePacks); err != nil {
-		return err
-	}
+		// Call RebuildIndex: rebuilds the index from the already loaded in-memory index.
+		// TODO in RebuildIndex: - Parallelize rebuilding
+		obsoleteIndexes, err := (repo.Index()).(*repository.MasterIndex).
```

Comment:
Could Repository.Index return *repository.MasterIndex instead, or would that trigger an import cycle?

Reply:
This is not that easy.. I dont't think that `internal/restic` should import `repostory` (despite of whether it might trigger an import cycle).
I tried to add `Obsolete()` and `RebuildIndex()` to the `Index` interface, but then some test structs need to be extended - I'll leave this as it is ATM.